File: nexus/frontend/src/types/events.ts
// ── Event types — must match backend/models/events.py EventType enum ──────────

export type EventType =
  | 'planner.start'
  | 'planner.done'
  | 'worker.spawn'
  | 'worker.command'
  | 'worker.result'
  | 'worker.failed'
  | 'worker.timeout'
  | 'worker.killed'
  | 'manager.verify'
  | 'manager.verify_ok'
  | 'manager.verify_fail'
  | 'manager.relevance_filter'
  | 'manager.package'
  | 'main_llm.start'
  | 'main_llm.stream'
  | 'main_llm.done'
  | 'token.update'
  | 'system.error'
  | 'session.done'
  | 'benchmark.started'
  | 'benchmark.progress'
  | 'benchmark.stream'
  | 'benchmark.complete'
  | 'benchmark.timeout'
  | 'benchmark.error'

export interface NexusEvent {
  type: EventType
  session_id: string
  timestamp: number          // Unix epoch float (seconds)
  data: Record<string, unknown>
}

// ── Token stats — mirrors backend/models/schemas.py TokenStats ────────────────
// Note: total_nexus and reduction_pct are @property on the Python side and
// NOT in model_dump(). Compute them here with helpers.

export interface TokenStats {
  planner_in: number
  planner_out: number
  worker_in: number
  worker_out: number
  manager_in: number
  manager_out: number
  main_in: number
  main_out: number
  naive_estimate: number
}

export const DEFAULT_STATS: TokenStats = {
  planner_in: 0, planner_out: 0,
  worker_in: 0, worker_out: 0,
  manager_in: 0, manager_out: 0,
  main_in: 0, main_out: 0,
  naive_estimate: 0,
}

export function totalNexus(s: TokenStats): number {
  return (
    s.planner_in + s.planner_out +
    s.worker_in + s.worker_out +
    s.manager_in + s.manager_out +
    s.main_in + s.main_out
  )
}

export function reductionPct(s: TokenStats): number {
  if (s.naive_estimate === 0) return 0
  return Math.max(0, (1 - totalNexus(s) / s.naive_estimate) * 100)
}

export function fmtTokens(n: number): string {
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`
  if (n >= 1_000) return `${(n / 1_000).toFixed(1)}K`
  return n.toLocaleString()
}

// ── Typed data accessors for specific events ──────────────────────────────────

export interface PlannerDoneData { task_count: number; tok_in: number; tok_out: number }
export interface WorkerSpawnData  { task_id: string; description: string; command: string }
export interface WorkerCommandData { task_id: string; command: string; output_lines: number }
export interface WorkerResultData  { task_id: string; citations: string[]; tok_in: number; tok_out: number }
export interface WorkerFailedData  { task_id: string; error: string }
export interface ManagerVerifyData { total_citations: number }
export interface ManagerVerifyOkData { citation: string }
export interface ManagerVerifyFailData { citation: string; reason: string }
export interface ManagerPackageData { verified: number; dropped: number; token_estimate: number }
export interface MainLlmStartData  { context_facts: number; context_tokens: number }
export interface MainLlmStreamData { chunk: string }
export interface MainLlmDoneData   { response: string; tokens: TokenStats }
export interface SessionDoneData   { tokens: TokenStats; nexus_total: number; naive_estimate: number; reduction_pct: number }
export interface SystemErrorData   { error: string; trace?: string }

export interface BenchmarkResult {
  tokens: number
  toolCalls: number
  timeSeconds: number
  model?: string
}

File: nexus/backend/models/events.py
import time
from enum import Enum
from typing import Any

from pydantic import BaseModel, model_validator


class EventType(str, Enum):
    # Planner
    PLANNER_START = "planner.start"
    PLANNER_DONE = "planner.done"

    # Worker
    WORKER_SPAWN = "worker.spawn"
    WORKER_COMMAND = "worker.command"
    WORKER_RESULT = "worker.result"
    WORKER_FAILED = "worker.failed"
    WORKER_TIMEOUT = "worker.timeout"
    WORKER_KILLED = "worker.killed"

    # Manager
    MANAGER_VERIFY = "manager.verify"
    MANAGER_VERIFY_OK = "manager.verify_ok"
    MANAGER_VERIFY_FAIL = "manager.verify_fail"
    MANAGER_RELEVANCE_FILTER = "manager.relevance_filter"
    MANAGER_PACKAGE = "manager.package"

    # Main LLM
    MAIN_LLM_START = "main_llm.start"
    MAIN_LLM_STREAM = "main_llm.stream"
    MAIN_LLM_DONE = "main_llm.done"

    # Token tracking
    TOKEN_UPDATE = "token.update"

    # Benchmark (Claude Code CLI race)
    BENCHMARK_STARTED  = "benchmark.started"
    BENCHMARK_PROGRESS = "benchmark.progress"
    BENCHMARK_STREAM   = "benchmark.stream"
    BENCHMARK_COMPLETE = "benchmark.complete"
    BENCHMARK_TIMEOUT  = "benchmark.timeout"
    BENCHMARK_ERROR    = "benchmark.error"

    # System
    ERROR = "system.error"
    SESSION_DONE = "session.done"


class NexusEvent(BaseModel):
    type: EventType
    session_id: str
    timestamp: float = 0.0
    data: dict[str, Any] = {}

    @model_validator(mode="after")
    def set_timestamp(self) -> "NexusEvent":
        if self.timestamp == 0.0:
            self.timestamp = time.time()
        return self
